lib/expand/phases/wait.go

```diff
@@ -0,0 +1,160 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"context"
+
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/fsm"
+	"github.com/gravitational/gravity/lib/httplib"
+	kubeutils "github.com/gravitational/gravity/lib/kubernetes"
+	"github.com/gravitational/gravity/lib/ops"
+	"github.com/gravitational/gravity/lib/status"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+)
+
+// NewWaitPlanet returns executor that waits for planet to start on the joining node
+func NewWaitPlanet(p fsm.ExecutorParams, operator ops.Operator) (*waitPlanetExecutor, error) {
+	logger := &fsm.Logger{
+		FieldLogger: logrus.WithFields(logrus.Fields{
+			constants.FieldPhase: p.Phase.ID,
+		}),
+		Key:      opKey(p.Plan),
+		Operator: operator,
+		Server:   p.Phase.Data.Server,
+	}
+	return &waitPlanetExecutor{
+		FieldLogger:    logger,
+		ExecutorParams: p,
+	}, nil
+}
+
+type waitPlanetExecutor struct {
+	// FieldLogger is used for logging
+	logrus.FieldLogger
+	// ExecutorParams is common executor params
+	fsm.ExecutorParams
+}
+
+// Execute executes the wait phase
+func (p *waitPlanetExecutor) Execute(ctx context.Context) error {
+	p.Progress.NextStep("Waiting for the planet to start")
+	p.Info("Waiting for the planet to start.")
+	err := utils.Retry(defaults.RetryInterval, defaults.RetryAttempts,
+		func() error {
+			planetStatus, err := status.FromPlanetAgent(ctx, nil)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			for _, nodeStatus := range planetStatus.Nodes {
+				if p.Phase.Data.Server.AdvertiseIP == nodeStatus.AdvertiseIP {
+					if nodeStatus.Status == status.NodeHealthy {
+						return nil
+					}
+				}
+			}
+			return trace.BadParameter("planet is not running yet: %#v",
+				planetStatus)
+		})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	p.Info("Planet is running.")
+	return nil
+}
+
+// Rollback is no-op for this phase
+func (*waitPlanetExecutor) Rollback(ctx context.Context) error {
+	return nil
+}
+
+// PreCheck is no-op for this phase
+func (*waitPlanetExecutor) PreCheck(ctx context.Context) error {
+	return nil
+}
+
+// PostCheck is no-op for this phase
+func (*waitPlanetExecutor) PostCheck(ctx context.Context) error {
+	return nil
+}
+
+// NewWaitK8s returns executor that waits for Kubernetes node to register
+func NewWaitK8s(p fsm.ExecutorParams, operator ops.Operator) (*waitK8sExecutor, error) {
+	client, err := httplib.GetUnprivilegedKubeClient() // Kubernetes should be up by now
```

Comment:
You should retry to avoid/minimize the "should" part.

Reply:
There's nothing really to retry here - this just creates the client. The phase execution itself it wrapped into retry.